src/chainparams.cpp

```diff
@@ -382,11 +381,39 @@ class SigNetParams : public CChainParams {
     }
 };
 
+std::unique_ptr<const CChainParams> CreateSignetChainParams(const SigNetOptions& options)
+{
+    return std::make_unique<const SigNetParams>(options);
+}
+
+std::unique_ptr<const CChainParams> CreateSignetChainParams(const ArgsManager& args)
+{
+    std::vector<std::string> seeds{};
+    if (args.IsArgSet("-signetseednode")) {
+        seeds = args.GetArgs("-signetseednode");
+    }
+    if (args.IsArgSet("-signetchallenge")) {
+        const auto signet_challenge = args.GetArgs("-signetchallenge");
+        if (signet_challenge.size() != 1) {
+            throw std::runtime_error(strprintf("%s: -signetchallenge cannot be multiple values.", __func__));
+        }
+        return CreateSignetChainParams(SigNetOptions{ParseHex(signet_challenge[0]), seeds});
+    } else {
+        auto opts = GetDefaultSignetOptions();
+        if (!seeds.empty()) {
```

Comment:
> I think that is taking it a bit far. 

Right I'm just talking about a minor change to the `SigNetOptions` struct that would simplify the changes here.

The only thing this PR really needs to do is get rid of `ArgsManager` usage in this part of the code, replacing `ArgsManager` lookups with `SigNetOptions` member accesses. I suggested a slightly different version of the `SigNetOptions` struct above that would let it provide the same information `ArgsManager` provides, in a 1:1 way so less code has to change, and so the struct is easier for callers to fill and customize.

> But if that were really the case, why not just pass a fully fleshed out const CChainParams& to the kernel in the first place?

This is fine, and I think it basically already happens later after `CChainParams` are initialized. I'm just looking at the earlier code used to initialize `CChainParams`. I do think alternate approaches are possible. Like instead of `SigNetOptions` being used to initialize `CChainParams` and have  `CChainParams` contain dynamic settings determined at runtime, `CChainParams` could just contain static hardcoded settings, and code that uses `CChainParams` could use combine it with runtime settings. This already happens with chain parameters like `nMinimumChainWork` that can be overridden on the command line.

> In that case, kernel/chainparams.h could be a headers only module, with the mainnet/signet/etc setup code staying in chainparams.cpp. I'm not sure if I'm missing a design goal here though.

I'm not sure exactly what this would look like but it does sound fine. Maybe similar to what I was describing above making `CChainParams` not vary depending on runtime options.

These alternate approaches are just bigger than changes I was suggesting above. All I was suggesting above was replacing ArgsManager uses with SigNetOptions in a 1:1 way instead of doing more complicated mappings and creating another set of defaults and making it harder to initialize.

Reply:
> I think that is taking it a bit far.

Yeah -- I think I agree, I just don't know why I agree? I was more hoping to understand why that would be taking it a bit far?

Reply:
Rereading I think we might be talking past each other. You mentioned things like "translating to text" which I'm not suggesting. I'm literally just agreeing with your initial suggestion `if (options.challenge.empty()) { /* use defaults */ } else { /* use values from options */ }` and saying how you can tweak the options struct definition so the same approach works for seeds as well, and the options struct does not need to contain any new or different information than what ArgsManager provides.

Later you mentioned passing "a fully fleshed out const CChainParams&" and having a "headers only module" and I think these changes sound ok, but maybe involve getting rid of the options struct or not using it here, and I'm just proposing tweaking the struct a little.

Reply:
@ryanofsky Yeah, I understand what you were saying, and tweaking the approach in this PR feels sensible to me too; I just don't feel like I intellectually understand the tradeoffs here, so was hoping thinking about more extreme alternatives might be revealing.

Three approaches:
 * `kernel::RegTestOptions` just contains the exact results from gArgs (so a vector of strings, not a map of enums etc)
 * what's implemented in this PR, with some tweaks
 * keep all the mainnet/testnet/etc defaults out of kernel, and just give it a prefilled CChainParams

I think this PR is better than the first option, because text is annoying, and the tweaks make it even better. I don't have a handle on the tradeoffs for the third approach, though.

Reply:
I think there is still a misunderstanding. I'm _not_ suggesting using text instead of enums or using hex strings instead of binary strings in the Options struct.

So I don't think I'm suggesting any of the approaches you listed above. I'm suggesting a fourth approach: `kernel::RegTestOptions` contains a 1:1 representation of information in `gArgs`. If a gArgs option is unset, corresponding Options parameter is unset or empty. If gArgs option is valid, corresponding options member contains the parsed option value. No special calls are needed to fill the options struct with defaults, because it just does the default thing by default and can be easily customized if desired. `bitcoind` behavior is unchanged and less code has to change because the options struct can be used as a straightforward replacement for ArgsManager calls.

I think if you want to go further and remove more code from the kernel, it could be good or bad, and the tradeoffs you are thinking about come into play. But I think the only thing we actually want to remove right now is ArgsManager. I don't think we need to something like removing preset chain parameters.

Reply:
Right -- I'm describing your (and my) suggestions as "some tweaks" to this PR. (They seem like obviously good ideas to me, so I was glossing over them in favour of the things I'm confused about)

Reply:
Makes sense!